### PR TITLE
small documentation tweaks based on MBTA practices

### PR DESCRIPTION
**Pull request by paulswartz**
**Best Practice Proposal:**

- use HTTP caching headers (producers and consumers) to avoid stale GTFS-RT data (rather than sticky sessions)
- `stop_time_updates` can be in the recent past to reflect a vehicle serving a trip earlier than scheduled
 
**Use Cases:** 

MBTA (@mbta) feeds use both of these practices. 
Swiftly's GTFS-RT feeds use both of these practices.

**Data Examples:**

https://cdn.mbta.com/realtime/TripUpdates.pb
https://cdn.mbta.com/realtime/VehiclePositions.pb

### CODE REVIEW

File: en/FeedHeader.md [hunk 1]
```diff
@@ -7,4 +7,4 @@
...
+|  | This timestamp value should always change if the feed contents change - the feed contents should not change without updating the header `timestamp`.<br>*Common mistakes* - If there are multiple instances of GTFS Realtime feed behind a load balancer, each instance may be pulling information from the realtime data source and publishing it to consumers slightly out of sync. If a GTFS Realtime consumer makes two back-to-back requests, and each request is served by a different GTFS Realtime feed instance, the same feed contents could potentially be returned to the consumer with different timestamps.<br>*Possible solution* - Producers should provide `Last-Modified` or `Etag` HTTP headers, and consumers should pass a `If-Modified-Since` or `If-None-Match` HTTP header to avoid receiving stale data. |
```
Comment (barbeau) [inline en/FeedHeader.md]: Good suggestion! IIRC in Tampa this wasn't possible though for some reason related to internal infrastructure, so the "sticky routes" solution was used instead. Maybe we should list them both as "possible solutions", with the HTTP header solution listed first?

File: en/TripUpdate.md [hunk 2]
```diff
@@ -11,6 +11,6 @@ General guidelines for trip cancellations:
...
 | stop_time_update | `stop_time_updates` for a given `trip_id` should be strictly ordered by increasing `stop_sequence` and no `stop_sequence` should be repeated. |
-|  | All TripUpdates should include at least one `stop_time_update` with a predicted arrival or departure time in the future. If all `stop_time_updates` for a trip reference past arrival and departure times, consumers should assume that no real-time data is available for the trip. |
+|  | All TripUpdates should include at least one `stop_time_update` with a predicted arrival or departure time in the future or recent past. If all `stop_time_updates` for a trip reference past arrival and departure times, consumers should assume that no real-time data is available for the trip. `stop_time_updates` in the recent past can be used to notify consumers that a vehicle has served the trip earlier than expected. |
```
Comment (barbeau) [inline en/TripUpdate.md]: I 100% agree with the last sentence (and the spec actually requires it of producers).

For the first sentence, could we change it to:

>All TripUpdates should include at least one `stop_time_update` with a predicted arrival or departure time in the future while the trip is in progress

This sentence was intended to address scenarios I've seen where the stop_time_updates trail behind the current time and presumably are actually observed schedule deviation values.

So if it's 10:00am, you'd see a header timestamp of 9:55am and stop_time_update values like:
* stop 1 - 9:50am
* stop 2 - 9:55am

and then at 10:05am you'd see a header timestamp of 10am and:
* stop 2 - 9:55am
* stop 3 - 10am

Related - I'd like to see TU.stop_time_updates be clarified to be only predictions, and TU.delay to be deprecated and replaced with VP.schedule_deviation, which would be for observed deviation from the schedule (not predictions).

**Review by barbeau — COMMENTED** [review]
@paulswartz thanks for the feedback! A few comments in-line.

File: en/FeedHeader.md (hunk 1 above)
Comment (derhuerst) [reply en/FeedHeader.md]: my 2 cents: 👍 for the change to leveraging HTTP caching. To be honest, I don't see why this (bad) internal design of a load balancing system should lead the GTFS-RT Best Practices to disregard well-known and commonly used HTTP semantics. 😉

Comment (barbeau) [reply en/FeedHeader.md]: If there is a consensus that sticky routes here is "bad design" and shouldn't be used in practice, and there isn't anything blocking use of HTTP headers in the majority of deployments, I don't object to removing the mention of sticky routes here.

Comment (derhuerst) [reply en/FeedHeader.md]: Okay, my post is a bit misleading, sorry. I don't see a reason not to use sticky routes internally, but no matter how the load balancing system is built, doing requests from the outside, it should act consistently: I shouldn't *ever* receive an older GTFS-RT feed on a subsequent request, right?

Or to rephrase it again: In this case, using "sticky routes" is a bad and insufficient solution to a *different* problem: one of internal consistency in the publisher's system. What if I request the GTFS-RT dump from 2 different IP addresses but process it in one place? In this case my system is *one* consumer but the load balancer has no way to guarantee a consistent behaviour.

Comment (barbeau) [reply en/FeedHeader.md]: >I shouldn't ever receive an older GTFS-RT feed on a subsequent request, right?

That's ideal, yes.

>What if I request the GTFS-RT dump from 2 different IP addresses but process it in one place? In this case my system is one consumer but the load balancer has no way to guarantee a consistent behaviour.

Yes, this is true. 

So maybe a better option is to further elaborate here on the desired behavior (consistency from the consumer's perspective) and how both of these options address (or don't address) the problem? 

In my experience often agencies don't have the in-house expertise to understand the pros/cons of their decisions on consumers.

Comment (derhuerst) [reply en/FeedHeader.md]: > So maybe a better option is to further elaborate here on the desired behavior (consistency from the consumer's perspective) and how both of these options address (or don't address) the problem?
> 
> In my experience often agencies don't have the in-house expertise to understand the pros/cons of their decisions on consumers.

I agree. But while using `Last-Modified` & `ETag` headers requires technical expertise (configuring a web server like nginx properly), it is a lot harder to get wrong than configuring a load balancer to use sticky routes *only*.

We could rephrase it to say that:

- `Last-Modified` & `ETag` headers are recommended. In combination with [conditional requests](https://developer.mozilla.org/en-US/docs/Web/HTTP/Conditional_requests), they are the intended way of *correctly* (and efficiently) specifying that or checking if a resource has changed.
- *If* agencies want to use a load balancer, they should make sure that all servers behind it have a consistent state, e.g. by synchronising the generated GTFS-RT data.

---

Related to the topic:

Configuring a single caching (reverse) proxy to

- periodically re-fetch the original/actual GTFS-RT feed when there's a request (e.g. every few seconds), *and*
- serve *all* requests from the cache

is a semantically correct and probably the most efficient setup. AFAIK most popular HTTP (reverse) proxies can be configured to do that, and having >1 GTFS-RT-generating instances shouldn't be necessary anymore, as it is guaranteed not to receive spikes of requests.

File: en/TripUpdate.md (hunk 2 above)
Comment (paulswartz) [reply en/TripUpdate.md]: > All TripUpdates should include at least one `stop_time_update` with a predicted arrival or departure time in the future while the trip is in progress

@barbeau Even if the trip isn't currently in progress (ended early) we'd still want to keep it around if it was scheduled to be running. Maybe a parenthetical?

> All TripUpdates should include at least one `stop_time_update` with a predicted arrival or departure time in the future while the trip is in progress (or scheduled to be in progress).

File: en/FeedHeader.md (hunk 1 above)
Comment (paulswartz) [reply en/FeedHeader.md]: On second thought, `ETag` doesn't actually help in this case since you could flip between two cache endpoints with different data. I'll work on re-phrasing.

File: en/TripUpdate.md (hunk 2 above)
Comment (barbeau) [reply en/TripUpdate.md]: @paulswartz Yes, my proposed text definitely wasn't intended to exclude that case - your new proposed text works for me!

File: en/FeedHeader.md [hunk 3]
```diff
@@ -7,4 +7,4 @@
...
+|  | This timestamp value should always change if the feed contents change - the feed contents should not change without updating the header `timestamp`.<br>*Common mistakes* - If there are multiple instances of GTFS Realtime feed behind a load balancer, each instance may be pulling information from the realtime data source and publishing it to consumers slightly out of sync. If a GTFS Realtime consumer makes two back-to-back requests, and each request is served by a different GTFS Realtime feed instance, the same feed contents could potentially be returned to the consumer with different timestamps.<br>*Possible solution* - Producers should provide a `Last-Modified` HTTP header, and consumers should pass their most recent `If-Modified-Since` HTTP header to avoid receiving stale data.<br>*Possible solution* - If HTTP headers cannot be used, options such as stick sessions can be used to ensure that each consumer is routed to the same producer server. |
```
Comment (paulswartz) [inline en/FeedHeader.md]:
```suggestion
|  | This timestamp value should always change if the feed contents change - the feed contents should not change without updating the header `timestamp`.<br>*Common mistakes* - If there are multiple instances of GTFS Realtime feed behind a load balancer, each instance may be pulling information from the realtime data source and publishing it to consumers slightly out of sync. If a GTFS Realtime consumer makes two back-to-back requests, and each request is served by a different GTFS Realtime feed instance, the same feed contents could potentially be returned to the consumer with different timestamps.<br>*Possible solution* - Producers should provide a `Last-Modified` HTTP header, and consumers should pass their most recent `If-Modified-Since` HTTP header to avoid receiving stale data.<br>*Possible solution* - If HTTP headers cannot be used, options such as sticky sessions can be used to ensure that each consumer is routed to the same producer server. |
```

**Comment by paulswartz** [comment]
One typo, but otherwise I think the new language is good!

File: en/TripUpdate.md [hunk 4]
```diff
@@ -11,6 +11,6 @@ General guidelines for trip cancellations:
...
 |  | If separate `VehiclePosition` and `TripUpdate` feeds are provided, [TripDescriptor](#TripDescriptor) and [VehicleDescriptor](#VehicleDescriptor) ID values pairing should match between the two feeds.<br/>For example, a `VehiclePosition` entity has `vehicle_id:A` and `trip_id:4`, then the corresponding `TripUpdate` entity should also have `vehicle_id:A` and `trip_id:4`. If any `TripUpdate` entity has `trip_id:4` and any `vehicle_id` other than 4, this is an error. |
 | stop_time_update | `stop_time_updates` for a given `trip_id` should be strictly ordered by increasing `stop_sequence` and no `stop_sequence` should be repeated. |
-|  | All TripUpdates should include at least one `stop_time_update` with a predicted arrival or departure time in the future. If all `stop_time_updates` for a trip reference past arrival and departure times, consumers should assume that no real-time data is available for the trip. |
+|  | All TripUpdates should include at least one stop_time_update with a predicted arrival or departure time in the future (or recent past) while the trip is in progress (or scheduled to be in progress). |
```
Comment (barbeau) [inline en/TripUpdate.md]: @paulswartz Coming back and looking at this with fresh eyes, I don't think the this new language is clear enough to avoid the case I mentioned in https://github.com/MobilityData/GTFS_Realtime_Best-Practices/pull/15#discussion_r644859538 (where all stop_time_updates are consistently in the past).

Could we reword this as:

>While the trip is in progress, all TripUpdates should include at least one stop_time_update with a predicted arrival or departure time in the future. Note that the [GTFS Realtime spec](https://github.com/google/transit/blob/master/gtfs-realtime/spec/en/trip-updates.md#stop-time-updates) says that producers should not drop a past StopTimeUpdate if it refers to a stop with a scheduled arrival time in the future for the given trip (i.e. the vehicle has passed the stop ahead of schedule), as otherwise it will be concluded that there is no update for this stop.

Technically the "SHOULD not drop" should be "MUST not drop" (and ideally "should include" should be "MUST include"), but that should be cleaned up in the RFC2119 update of the GTFS Realtime spec.

cc @scmcca

Comment (paulswartz) [reply en/TripUpdate.md]: @barbeau I don't think that's quite right. In the case where a vehicle gets to the final stop early, consumers should continue to provide the TripUpdate until the *scheduled* times are in the past, not only the StopTimeUpdates.

Comment (barbeau) [reply en/TripUpdate.md]: @paulswartz Ah, good point - and that language is copy/paste from the spec, so it should change too.

What if we remove the word "past" from "past StopTimeUpdate" (both here and in the spec)?

>While the trip is in progress, all TripUpdates should include at least one stop_time_update with a predicted arrival or departure time in the future. Note that the [GTFS Realtime spec](https://github.com/google/transit/blob/master/gtfs-realtime/spec/en/trip-updates.md#stop-time-updates) says that producers should not drop a StopTimeUpdate if it refers to a stop with a scheduled arrival time in the future for the given trip (i.e. the vehicle has passed the stop ahead of schedule), as otherwise it will be concluded that there is no update for this stop.

Comment (paulswartz) [reply en/TripUpdate.md]: @barbeau sorry for the late reply, but that language LGTM

**Comment by sccmcca** [comment]
Looks like the consensus is clear. I've applied the changes agreed on above and will go ahead and merge. Thanks!
